auth/range_perm_cache.go

```diff
@@ -146,18 +87,44 @@ func checkKeyPerm(cachedPerms *unifiedRangePermissions, key, rangeEnd []byte, pe
 		plog.Panicf("unknown auth type: %v", permtyp)
 	}
 
-	requiredPerm := &rangePerm{begin: key, end: rangeEnd}
+	ivl := adt.NewStringAffineInterval(key, rangeEnd)
 
-	for _, perm := range tocheck {
-		if isSubset(requiredPerm, perm) {
-			return true
+	minBegin := adt.StringComparable("")
+	maxEnd := adt.StringComparable("\xff")
+
+	tocheck.Visit(ivl, func(n *adt.IntervalValue) bool {
+		if n.Ivl.Begin.Compare(minBegin) < 0 {
+			minBegin = n.Ivl.Begin.(adt.StringComparable)
+		}
+
+		if maxEnd.Compare(n.Ivl.End) < 0 {
+			maxEnd = n.Ivl.End.(adt.StringComparable)
 		}
+
```

Comment:
I missed the case... thanks! Possibly
1. visit every interval that overlaps with the required range
1. check all the intervals are continuous
1. if they are not continuous, permission denied
1. `return minBegin.Compare(adt.StringComparable(key)) < 0 && maxEnd.Compare(adt.StringComparable(maxEnd)) < 0`

I'll also add some test cases. At first I thought that the implementation of interval tree has its own tests, so adding new tests would be redundant. But clearly it was wrong...